### ISSUE TITLE
Segmentation fault with sitk.ImageSeriesReader() GetGDCMSeriesIDs()

### ISSUE THREAD

**Issue by guatavita** [SimpleITK/SimpleITK#1466]
Hello, 

I had to run DICOM readers on thousands of folder that can have multiple series per folder. I noticed that python will randomly throw a segmentation fault when calling GetGDCMSeriesIDs() inside multiple threads.

The segmentation fault was isolated on that specific command line. This happened randomly after scanning the unique series ID in 1000-8000 folders I am not sure if that command line can have any memory management issue when called thousands of times via different thread, while keeping the reader local to the thread of course. Is that something that should be acknowledge in the limitation of the sitk.ImageSeriesReader() ? 

The segmentation fault happened in Windows 10 and Linux (Unbuntu Docker) using python virtual environment tested with SimpleITK 2.0.2 and 2.1.1

I can work on a simple code example if required.

Best,

**Comment by dave3d** [comment]
A code example would be very helpful.  I don't think the GDCM reader is thread safe, though, so that could be the cause.

**Comment by guatavita** [comment]
Alright, 

Here is an example. Hope this help!

```
import os
import glob
from queue import Queue
from threading import Thread
from multiprocessing import cpu_count
import SimpleITK as sitk


class Dicom_Reader(object):
    def __init__(self, input_dir, nb_threads=int(0.9 * cpu_count())):
        self.input_dir = input_dir
        self.folders_with_dcm = []
        self.nb_threads = nb_threads
        self.walk_root_folder()
        self.read_dicom_series()

    def walk_root_folder(self):
        for root, dirs, files in os.walk(self.input_dir, topdown=False):
            if glob.glob(os.path.join(root, '*.dcm')):
                self.folders_with_dcm.append(root)

    def dicom_reader_worker(self, A):
        q = A[0]
        while True:
            item = q.get()
            if item is None:
                break
            else:
                it, dicom_folder = item
                reader = sitk.ImageSeriesReader()
                reader.SetGlobalWarningDisplay(False)
                print('running {} {}'.format(it, dicom_folder))
                try:
                    series_ids_list = reader.GetGDCMSeriesIDs(dicom_folder)
                except:
                    print('Failed on {}'.format(dicom_folder))
            q.task_done()

    def read_dicom_series(self):
        q = Queue(maxsize=self.nb_threads)
        A = (q,)
        threads = []
        for worker in range(self.nb_threads):
            t = Thread(target=self.dicom_reader_worker, args=(A,))
            t.start()
            threads.append(t)

        for it, dicom_folder in enumerate(self.folders_with_dcm):
            item = [it, dicom_folder]
            q.put(item)

        for worker in range(self.nb_threads):
            q.put(None)
        for t in threads:
            t.join()


def main():
    input_dir = r'PATH_TO_DICOM_DIRECTORY'
    dicom_explorer = Dicom_Reader(input_dir=input_dir)


if __name__ == '__main__':
    main()
```

To solve this random segmentation fault, I switched to pydicom to extract the unique series ID in a dicom folder using any of these two functions (depending if you want to return the filenames of the seriesID).

```
def get_unique_series_ids(dicom_folder):
    '''
    :param dicom_folder:
    :return: list of unique seriesID in the folder
    '''
    series_id_list = []
    filenames = glob.glob(os.path.join(dicom_folder, "*.dcm"))
    for filename in filenames:
        try:
            ds = pydicom.read_file(filename)
        except:
            continue
        series_id = ds.get('SeriesInstanceUID')
        if series_id not in series_id_list:
            series_id_list.append(series_id)
    return series_id_list


def get_unique_series_ids_filenames(dicom_folder):
    '''
    :param dicom_folder:
    :return: dictionary with list of .dcm files per seriesID
    '''
    series_id_filenames = {}
    filenames = glob.glob(os.path.join(dicom_folder, "*.dcm"))
    for filename in filenames:
        try:
            ds = pydicom.read_file(filename)
        except:
            continue
        series_id = ds.get('SeriesInstanceUID')
        if series_id not in series_id_filenames:
            series_id_filenames[series_id] = []
        series_id_filenames[series_id].append(filename)
    return series_id_filenames
```

Best,

**Comment by zivy** [comment]
Hello @guatavita,

Your code appears to assume that the DICOM files have a `dcm` extension, this often isn't the case. A more generic approach is implement in the [code available here](https://github.com/InsightSoftwareConsortium/SimpleITK-Notebooks/blob/b0f275b3d6c77a92f5733434e29f9608d72c9749/Python/characterize_data.py#L238-L345). 

It traverses the directory structure, identifies the DICOM files based on the file content, aggregates the file names from the same study-series into a list and then reads them. Note that files from the same study-series may be stored in different directories. Commonly happens when data is written to a CD in multiple directories and the files have the same name (i.e. dir1/Z0 and dir2/Z0 both files belong to the same series). The code I pointed to deals with this situation too.

Let us know if this approach worked around your issue.

**Comment by guatavita** [comment]
Thank you @zivy, that's some interesting suggestions and yes I assumed all dicom files should have .dcm extension.

I had to move away from sitk readers anyway. This was cause too many Segmentation faults in multi-threads (60 for example). I had to read, classify and convert ~500,000 dicom files that where in ~110,000 folders and pydicom was more robust to run that in python multi-threading. Plus, pydicom allowed me to read the dicom files without loading the pixel data, which was 5-10 times faster.

Hope this discussion helps localize some issue/limitations for others. I'll definitely rely on sitk for other processes though.

Best,
